File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
class Solution {
    static HashSet<Integer> hs = new HashSet<>();
    static boolean[] isPrime;
    public int solution(String numbers) {
        isPrime = new boolean[(int)Math.pow(10, numbers.length())];
        isPrime[0] = true;
        isPrime[1] = true;
        for(int i=2; i<=Math.sqrt(isPrime.length); i++){
            if(!isPrime[i]){
                for(int j=2; i*j<isPrime.length; j++){
                    isPrime[i*j] = true;
                }
            }
        }
        int[] numberArr = new int[numbers.length()];
        for(int i=0; i<numbers.length(); i++){
            numberArr[i] = numbers.charAt(i) - '0';
        }
        Arrays.sort(numberArr);
        for(int depth=1; depth<=numberArr.length; depth++){
            boolean[] visited = new boolean[numberArr.length];
            List<Integer> list = new ArrayList<>();
            getNumber(numberArr, visited, list, depth);
        }
        System.out.println(hs);
        int answer = hs.size();
        return answer;
    }
    public void getNumber(int[] numberArr, boolean[] visited, List<Integer> list, int depth){
        if(list.size() == depth){
            int result = 0;
            for(int i=0; i<list.size(); i++){
                result += Math.pow(10, i) * list.get(list.size()-1-i);
            }
            if(!isPrime[result]){
                hs.add(result);
            }
            return;
        }
        for(int i=0; i<numberArr.length; i++){
            if(visited[i]){
                continue;
            }
            visited[i] = true;
            list.add(numberArr[i]);
            getNumber(numberArr, visited, list, depth);
            visited[i] = false;
            list.remove(list.size()-1);
        }
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
import math
def solution(numbers):
    isPrime = [True] * 10000000
    isPrime[0] = isPrime[1] = False
    for i in range(int(math.sqrt(10000000))):
        if isPrime[i]:
            j = 2
            while i*j < 10000000:
                isPrime[i*j] = False
                j += 1
        
    def getNumber(numbers, visited, result, depth):
        if len(result) == depth:
            sum = 0
            for i in range(len(result)):
                sum += int(math.pow(10, i)) * result[len(result)-1-i]
            if isPrime[sum]:
                hs.add(sum)
            return
        for i in range(len(numbers)):
            if visited[i]:
                continue
            visited[i] = True
            result.append(int(numbers[i]))
            getNumber(numbers, visited, result, depth)
            visited[i] = False
            result.pop()
    
    hs = set()
    for depth in range(1, len(numbers)+1):
        visited = [False] * len(numbers)
        result = []
        getNumber(numbers, visited, result, depth)
            
    answer = len(hs)
    return answer